orderbook/src/account_balances.rs

```diff
@@ -0,0 +1,195 @@
+use anyhow::{Context, Result};
+use ethcontract::{batch::CallBatch, errors::MethodError, Http, Web3};
+use futures::future::{join, join_all, ok, FutureExt as _};
+use std::collections::HashMap;
+use std::sync::Mutex;
+
+use contracts::IERC20;
+use primitive_types::{H160, U256};
+
+#[async_trait::async_trait]
+pub trait BalanceFetching: Send + Sync {
+    // warm cache and setup updates in the background
+    async fn register(&self, owner: H160, token: H160) -> Result<()>;
+
+    // Returns the current balance available to the allowance manager for the given owner and token.
+    // Should be non-blocking
+    fn get_balance(&self, owner: H160, token: H160) -> U256;
+}
+
+pub struct Web3BalanceFetcher {
+    web3: Web3<Http>,
+    allowance_manager: H160,
+    // Mapping of address, token to balance, allowance
+    balances: Mutex<HashMap<SubscriptionKey, SubscriptionValue>>,
+}
+
+#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq)]
+struct SubscriptionKey {
+    owner: H160,
+    token: H160,
+}
+
+#[derive(Clone, Default)]
+struct SubscriptionValue {
+    balance: U256,
+    allowance: U256,
+}
+
+impl Web3BalanceFetcher {
+    pub fn new(web3: Web3<Http>, allowance_manager: H160) -> Self {
+        Self {
+            web3,
+            allowance_manager,
+            balances: Default::default(),
+        }
+    }
+
+    pub async fn update(&self) {
+        let subscriptions: Vec<_> = {
+            let map = self.balances.lock().expect("mutex holding thread panicked");
+            map.keys().cloned().collect()
+        };
+        let _ = self.register_many(subscriptions.iter()).await;
+    }
+
+    async fn register_many(
+        &self,
+        subscriptions: impl Iterator<Item = &SubscriptionKey>,
+    ) -> Result<()> {
+        let mut batch = CallBatch::new(self.web3.transport());
+        let calls = subscriptions
+            .map(|subscription| {
+                let instance = IERC20::at(&self.web3, subscription.token);
+                join(
+                    instance
+                        .balance_of(subscription.owner)
+                        .view()
+                        .batch_call(&mut batch),
+                    instance
+                        .allowance(subscription.owner, self.allowance_manager)
+                        .view()
+                        .batch_call(&mut batch),
+                )
+                .then(move |(balance, allowance)| {
+                    self.update_with(subscription, balance, allowance);
+                    ok::<_, MethodError>(())
```

Comment:
This means the mutex is locked for every individual request. We already get all the responses at the same time into a vector so it would be better to only lock the mutex once to apply all of them.